File: morse/practice.py
"""Practice mode: targets to key, scoring of the copy, and a rhythm report.

Pure logic shared with ``web/js/practice.js`` (same word list, same
generators, same scoring), so both apps grade the same way.

* :class:`Practice` hands out targets: common words, call signs, five-digit
  groups, or a mix.  Seeded, so tests and the two ports agree.
* :func:`score` aligns what was sent against the target with the edit
  distance and reports correct letters, errors and accuracy.
* :func:`rhythm` measures how far the operator's marks and gaps sit from the
  ideal 1 : 3 : 7 proportions, in percent of the intended element.
"""
from __future__ import annotations

import random
from dataclasses import dataclass
from typing import Literal, Sequence

from morse.runs import Run

__all__ = ["KINDS", "WORDS", "Practice", "Rhythm", "Score", "call_sign", "digit_group", "rhythm", "score"]

Kind = Literal["words", "calls", "digits", "mixed"]
KINDS: tuple[str, ...] = ("words", "calls", "digits", "mixed")

WORDS: tuple[str, ...] = (
    "THE", "AND", "FOR", "YOU", "ARE", "WITH", "THIS", "HAVE", "FROM", "THAT", "NOT", "BUT",
    "ALL", "CAN", "HER", "WAS", "ONE", "OUR", "OUT", "DAY", "GET", "HAS", "HIM", "HIS", "HOW",
    "MAN", "NEW", "NOW", "OLD", "SEE", "TWO", "WAY", "WHO", "BOY", "DID", "ITS", "LET", "PUT",
    "SAY", "SHE", "TOO", "USE", "GOOD", "TIME", "YEAR", "WORK", "BACK", "CALL", "COME", "EACH",
    "FIND", "GIVE", "HAND", "HELP", "HERE", "HOME", "JUST", "KEEP", "KIND", "KNOW", "LAST",
    "LIKE", "LINE", "LIVE", "LONG", "LOOK", "MAKE", "MANY", "MORE", "MOST", "MOVE", "MUCH",
    "MUST", "NAME", "NEED", "NEXT", "ONLY", "OPEN", "OVER", "PART", "PLAY", "READ", "REAL",
    "SAME", "SEND", "SHOW", "SIDE", "SOME", "TAKE", "TELL", "THAN", "THEM", "THEN", "THEY",
    "TURN", "VERY", "WANT", "WEEK", "WELL", "WHAT", "WHEN", "WILL", "WORD", "ABOUT", "AFTER",
    "AGAIN", "COULD", "EVERY", "FIRST", "GREAT", "HOUSE", "LARGE", "NEVER", "OTHER", "PLACE",
    "RADIO", "RIGHT", "SMALL", "SOUND", "STILL", "THEIR", "THERE", "THESE", "THING", "THINK",
    "THREE", "UNDER", "WATER", "WHERE", "WHICH", "WORLD", "WOULD", "WRITE", "MORSE", "SIGNAL",
    "BEEPER", "HELLO", "PARIS", "CODEX", "ANTENNA", "REPEAT", "READY", "AGAIN", "OVER", "OUT",
)

_LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
_PREFIXES = ("K", "W", "N", "AA", "KB", "KC", "KD", "WA", "WB", "VE", "VK", "G", "M", "DL", "JA", "F", "EA", "PY")


def call_sign(rng: random.Random) -> str:
    """A plausible amateur call sign: prefix, one digit, one to three letters."""
    prefix = rng.choice(_PREFIXES)
    suffix = "".join(rng.choice(_LETTERS) for _ in range(rng.randint(1, 3)))
    return f"{prefix}{rng.randint(0, 9)}{suffix}"


def digit_group(rng: random.Random, length: int = 5) -> str:
    """A group of digits, the classic numbers drill."""
    return "".join(str(rng.randint(0, 9)) for _ in range(length))


class Practice:
    """Hands out targets of one kind; ``seed`` makes the sequence reproducible."""

    def __init__(self, kind: Kind = "words", seed: int | None = None) -> None:
        if kind not in KINDS:
            raise ValueError(f"kind must be one of {KINDS}, got {kind!r}")
        self.kind: Kind = kind
        self._rng = random.Random(seed)
        self.target: str | None = None
        self.asked: int = 0
        self.correct: int = 0

    def set_kind(self, kind: Kind) -> None:
        if kind not in KINDS:
            raise ValueError(f"kind must be one of {KINDS}, got {kind!r}")
        self.kind = kind

    def next_target(self) -> str:
        """Draw the next target (never the same as the current one)."""
        for _ in range(20):
            kind = self.kind
            if kind == "mixed":
                kind = self._rng.choice(("words", "words", "calls", "digits"))
            if kind == "words":
                target = self._rng.choice(WORDS)
            elif kind == "calls":
                target = call_sign(self._rng)
            else:
                target = digit_group(self._rng)
            if target != self.target:
                break
        self.target = target
        self.asked += 1
        return target

    def record(self, result: Score) -> None:
        """Count a checked target toward the session tally."""
        if result.correct == result.total and result.errors == 0:
            self.correct += 1

    def reset_session(self) -> None:
        self.asked = 0
        self.correct = 0
        self.target = None


@dataclass(frozen=True)
class Score:
    """How a copy compares with its target."""

    target: str
    sent: str
    correct: int
    """Target characters reproduced in order (the edit-distance alignment's matches)."""
    errors: int
    """Edit distance: substitutions, insertions and deletions."""
    total: int
    """Characters in the target."""

    @property
    def accuracy(self) -> float:
        """1.0 for a perfect copy; errors beyond the target's length drive it to 0."""
        if self.total == 0:
            return 1.0 if not self.sent else 0.0
        return max(0.0, 1.0 - self.errors / self.total)

    @property
    def perfect(self) -> bool:
        return self.errors == 0 and self.total > 0


def _normalise(text: str) -> str:
    return " ".join(text.upper().split())


def score(target: str, sent: str) -> Score:
    """Grade ``sent`` against ``target`` (case and surplus whitespace ignored)."""
    a = _normalise(target)
    b = _normalise(sent)
    n, m = len(a), len(b)
    # Levenshtein distance with a match count recovered from the alignment.
    dist = [[0] * (m + 1) for _ in range(n + 1)]
    for i in range(1, n + 1):
        dist[i][0] = i
    for j in range(1, m + 1):
        dist[0][j] = j
    for i in range(1, n + 1):
        for j in range(1, m + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            dist[i][j] = min(dist[i - 1][j] + 1, dist[i][j - 1] + 1, dist[i - 1][j - 1] + cost)
    # Walk back to count matched characters.
    i, j, matches = n, m, 0
    while i > 0 and j > 0:
        if a[i - 1] == b[j - 1] and dist[i][j] == dist[i - 1][j - 1]:
            matches += 1
            i -= 1
            j -= 1
        elif dist[i][j] == dist[i - 1][j - 1] + 1:
            i -= 1
            j -= 1
        elif dist[i][j] == dist[i - 1][j] + 1:
            i -= 1
        else:
            j -= 1
    return Score(target=a, sent=b, correct=matches, errors=dist[n][m], total=n)


@dataclass(frozen=True)
class Rhythm:
    """Timing quality of a keyed passage, in percent of the intended element length."""

    mark_error_pct: float
    """Mean |measured − ideal| / ideal over marks, ideal being 1 or 3 dits."""
    gap_error_pct: float
    """Mean |measured − ideal| / ideal over gaps, ideal being 1, 3 or 7 dits."""
    marks: int
    gaps: int

    @property
    def error_pct(self) -> float:
        """Marks and gaps together, weighted by count."""
        n = self.marks + self.gaps
        if n == 0:
            return 0.0
        return (self.mark_error_pct * self.marks + self.gap_error_pct * self.gaps) / n


def rhythm(runs: Sequence[Run], dit_ms: float) -> Rhythm:
    """Compare each run with the nearest ideal element at ``dit_ms``.

    Marks are dits (1) or dahs (3) by the 2-dit split; gaps are element (1),
    letter (3) or word (7) gaps by the 2- and 5-dit splits. Leading and
    trailing gaps are not runs, so pass only the runs between the first and
    last mark.
    """
    if not dit_ms > 0:
        raise ValueError(f"dit_ms must be positive, got {dit_ms!r}")
    mark_errors: list[float] = []
    gap_errors: list[float] = []
    for run in runs:
        units = run.ms / dit_ms
        if run.on:
            ideal = 1.0 if units < 2.0 else 3.0
            mark_errors.append(abs(units - ideal) / ideal)
        else:
            ideal = 1.0 if units < 2.0 else 3.0 if units < 5.0 else 7.0
            gap_errors.append(abs(units - ideal) / ideal)
    mean = lambda xs: (sum(xs) / len(xs)) if xs else 0.0  # noqa: E731
    return Rhythm(
        mark_error_pct=100.0 * mean(mark_errors),
        gap_error_pct=100.0 * mean(gap_errors),
        marks=len(mark_errors),
        gaps=len(gap_errors),
    )

File: web/js/practice.js
/**
 * Practice mode: targets to key, scoring of the copy, and a rhythm report.
 *
 * Port of `morse/practice.py`: same word list, same generators (driven by a
 * seeded PRNG so tests are reproducible), same edit-distance scoring and the
 * same rhythm measure, so both apps grade the same way. No DOM, no audio.
 */

export const KINDS = ["words", "calls", "digits", "mixed"];

export const WORDS = [
  "THE", "AND", "FOR", "YOU", "ARE", "WITH", "THIS", "HAVE", "FROM", "THAT", "NOT", "BUT",
  "ALL", "CAN", "HER", "WAS", "ONE", "OUR", "OUT", "DAY", "GET", "HAS", "HIM", "HIS", "HOW",
  "MAN", "NEW", "NOW", "OLD", "SEE", "TWO", "WAY", "WHO", "BOY", "DID", "ITS", "LET", "PUT",
  "SAY", "SHE", "TOO", "USE", "GOOD", "TIME", "YEAR", "WORK", "BACK", "CALL", "COME", "EACH",
  "FIND", "GIVE", "HAND", "HELP", "HERE", "HOME", "JUST", "KEEP", "KIND", "KNOW", "LAST",
  "LIKE", "LINE", "LIVE", "LONG", "LOOK", "MAKE", "MANY", "MORE", "MOST", "MOVE", "MUCH",
  "MUST", "NAME", "NEED", "NEXT", "ONLY", "OPEN", "OVER", "PART", "PLAY", "READ", "REAL",
  "SAME", "SEND", "SHOW", "SIDE", "SOME", "TAKE", "TELL", "THAN", "THEM", "THEN", "THEY",
  "TURN", "VERY", "WANT", "WEEK", "WELL", "WHAT", "WHEN", "WILL", "WORD", "ABOUT", "AFTER",
  "AGAIN", "COULD", "EVERY", "FIRST", "GREAT", "HOUSE", "LARGE", "NEVER", "OTHER", "PLACE",
  "RADIO", "RIGHT", "SMALL", "SOUND", "STILL", "THEIR", "THERE", "THESE", "THING", "THINK",
  "THREE", "UNDER", "WATER", "WHERE", "WHICH", "WORLD", "WOULD", "WRITE", "MORSE", "SIGNAL",
  "BEEPER", "HELLO", "PARIS", "CODEX", "ANTENNA", "REPEAT", "READY", "AGAIN", "OVER", "OUT",
];

const LETTERS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
const PREFIXES = ["K", "W", "N", "AA", "KB", "KC", "KD", "WA", "WB", "VE", "VK", "G", "M", "DL", "JA", "F", "EA", "PY"];

/**
 * Small seeded PRNG (mulberry32); `Math.random` when no seed is given.
 * @param {number | null | undefined} seed
 * @returns {() => number} uniform in [0, 1)
 */
export function makeRng(seed) {
  if (seed === null || seed === undefined) return Math.random;
  let a = (seed >>> 0) || 0x9e3779b9;
  return () => {
    a = (a + 0x6d2b79f5) | 0;
    let t = Math.imul(a ^ (a >>> 15), 1 | a);
    t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

const pick = (rng, items) => items[Math.floor(rng() * items.length)];
const randint = (rng, lo, hi) => lo + Math.floor(rng() * (hi - lo + 1));

/** A plausible amateur call sign: prefix, one digit, one to three letters. @param {() => number} rng */
export function callSign(rng) {
  const prefix = pick(rng, PREFIXES);
  const n = randint(rng, 1, 3);
  let suffix = "";
  for (let i = 0; i < n; i++) suffix += LETTERS[Math.floor(rng() * LETTERS.length)];
  return `${prefix}${randint(rng, 0, 9)}${suffix}`;
}

/** A group of digits, the classic numbers drill. @param {() => number} rng @param {number} [length=5] */
export function digitGroup(rng, length = 5) {
  let out = "";
  for (let i = 0; i < length; i++) out += String(randint(rng, 0, 9));
  return out;
}

export class Practice {
  /**
   * @param {"words" | "calls" | "digits" | "mixed"} [kind="words"]
   * @param {number | null} [seed=null]
   */
  constructor(kind = "words", seed = null) {
    if (!KINDS.includes(kind)) throw new RangeError(`kind must be one of ${KINDS.join(", ")}, got ${kind}`);
    this.kind = kind;
    this._rng = makeRng(seed);
    /** @type {string | null} */
    this.target = null;
    this.asked = 0;
    this.correct = 0;
  }

  /** @param {"words" | "calls" | "digits" | "mixed"} kind */
  setKind(kind) {
    if (!KINDS.includes(kind)) throw new RangeError(`kind must be one of ${KINDS.join(", ")}, got ${kind}`);
    this.kind = kind;
  }

  /** Draw the next target (never the same as the current one). @returns {string} */
  nextTarget() {
    let target = this.target ?? "";
    for (let i = 0; i < 20; i++) {
      let kind = this.kind;
      if (kind === "mixed") kind = pick(this._rng, ["words", "words", "calls", "digits"]);
      if (kind === "words") target = pick(this._rng, WORDS);
      else if (kind === "calls") target = callSign(this._rng);
      else target = digitGroup(this._rng);
      if (target !== this.target) break;
    }
    this.target = target;
    this.asked += 1;
    return target;
  }

  /** Count a checked target toward the session tally. @param {Score} result */
  record(result) {
    if (result.correct === result.total && result.errors === 0) this.correct += 1;
  }

  resetSession() {
    this.asked = 0;
    this.correct = 0;
    this.target = null;
  }
}

export class Score {
  constructor(target, sent, correct, errors, total) {
    this.target = target;
    this.sent = sent;
    /** Target characters reproduced in order (the alignment's matches). */
    this.correct = correct;
    /** Edit distance: substitutions, insertions and deletions. */
    this.errors = errors;
    /** Characters in the target. */
    this.total = total;
  }

  /** 1 for a perfect copy; errors beyond the target's length drive it to 0. @returns {number} */
  get accuracy() {
    if (this.total === 0) return this.sent ? 0 : 1;
    return Math.max(0, 1 - this.errors / this.total);
  }

  get perfect() {
    return this.errors === 0 && this.total > 0;
  }
}

const normalise = (text) => text.toUpperCase().split(/\s+/).filter(Boolean).join(" ");

/**
 * Grade `sent` against `target` (case and surplus whitespace ignored).
 * @param {string} target @param {string} sent @returns {Score}
 */
export function score(target, sent) {
  const a = normalise(target);
  const b = normalise(sent);
  const n = a.length;
  const m = b.length;
  const dist = Array.from({ length: n + 1 }, () => new Array(m + 1).fill(0));
  for (let i = 1; i <= n; i++) dist[i][0] = i;
  for (let j = 1; j <= m; j++) dist[0][j] = j;
  for (let i = 1; i <= n; i++) {
    for (let j = 1; j <= m; j++) {
      const cost = a[i - 1] === b[j - 1] ? 0 : 1;
      dist[i][j] = Math.min(dist[i - 1][j] + 1, dist[i][j - 1] + 1, dist[i - 1][j - 1] + cost);
    }
  }
  let i = n;
  let j = m;
  let matches = 0;
  while (i > 0 && j > 0) {
    if (a[i - 1] === b[j - 1] && dist[i][j] === dist[i - 1][j - 1]) {
      matches += 1;
      i -= 1;
      j -= 1;
    } else if (dist[i][j] === dist[i - 1][j - 1] + 1) {
      i -= 1;
      j -= 1;
    } else if (dist[i][j] === dist[i - 1][j] + 1) {
      i -= 1;
    } else {
      j -= 1;
    }
  }
  return new Score(a, b, matches, dist[n][m], n);
}

export class Rhythm {
  constructor(markErrorPct, gapErrorPct, marks, gaps) {
    this.markErrorPct = markErrorPct;
    this.gapErrorPct = gapErrorPct;
    this.marks = marks;
    this.gaps = gaps;
  }

  /** Marks and gaps together, weighted by count. @returns {number} */
  get errorPct() {
    const n = this.marks + this.gaps;
    if (n === 0) return 0;
    return (this.markErrorPct * this.marks + this.gapErrorPct * this.gaps) / n;
  }
}

/**
 * Compare each run with the nearest ideal element at `ditMs`: marks are dits
 * (1) or dahs (3) by the 2-dit split; gaps are element (1), letter (3) or
 * word (7) gaps by the 2- and 5-dit splits. Pass only the runs between the
 * first and last mark.
 * @param {Array<{on: boolean, ms: number}>} runs @param {number} ditMs @returns {Rhythm}
 */
export function rhythm(runs, ditMs) {
  if (!(ditMs > 0)) throw new RangeError(`ditMs must be positive, got ${ditMs}`);
  const markErrors = [];
  const gapErrors = [];
  for (const run of runs) {
    const units = run.ms / ditMs;
    if (run.on) {
      const ideal = units < 2 ? 1 : 3;
      markErrors.push(Math.abs(units - ideal) / ideal);
    } else {
      const ideal = units < 2 ? 1 : units < 5 ? 3 : 7;
      gapErrors.push(Math.abs(units - ideal) / ideal);
    }
  }
  const mean = (xs) => (xs.length ? xs.reduce((s, x) => s + x, 0) / xs.length : 0);
  return new Rhythm(100 * mean(markErrors), 100 * mean(gapErrors), markErrors.length, gapErrors.length);
}
